contracts/leaser/src/migrate.rs

```diff
@@ -2,66 +2,97 @@ use lease::api::MigrateMsg;
 
-pub fn migrate_leases<I>(leases: I, lease_code_id: u64) -> ContractResult<Batch>
+#[derive(Default)]
+#[cfg_attr(test, derive(Debug, Eq, PartialEq))]
+pub struct MigrationResult {
+    pub msgs: Batch,
+    /// None if the number of processed instances is less than the `max_leases`
+    pub last_instance: Option<Addr>,
+}
+
+pub fn migrate_leases<I>(
+    leases: I,
+    lease_code_id: u64,
+    max_leases: NbInstances,
+) -> ContractResult<MigrationResult>
 where
     I: Iterator<Item = ContractResult<Addr>>,
 {
-    let no_msgs = MigrateBatch::new(lease_code_id);
-    let migrated_msgs = leases.fold(no_msgs, MigrateBatch::migrate_lease);
+    let no_msgs = MigrateBatch::new(lease_code_id, max_leases);
+    let migrated_msgs = leases
+        .take(max_leases.try_into()?)
+        .fold(no_msgs, MigrateBatch::migrate_lease);
     migrated_msgs.try_into()
 }
 
+impl MigrationResult {
+    pub fn try_add_msgs<F>(mut self, add_fn: F) -> ContractResult<Self>
+    where
+        F: FnOnce(&mut Batch) -> ContractResult<()>,
+    {
+        add_fn(&mut self.msgs).map(|()| self)
+    }
+}
 struct MigrateBatch {
     new_code_id: u64,
-    may_batch: ContractResult<Batch>,
+    leases_left: NbInstances,
+    may_result: ContractResult<MigrationResult>,
 }
 impl MigrateBatch {
-    fn new(new_code_id: u64) -> Self {
+    fn new(new_code_id: u64, max_leases: NbInstances) -> Self {
         Self {
             new_code_id,
-            may_batch: Ok(Batch::default()),
+            leases_left: max_leases,
+            may_result: Ok(Default::default()),
         }
     }
 
     fn migrate_lease(mut self, lease_contract: ContractResult<Addr>) -> Self {
-        let op = |mut batch: Batch| {
+        let op = |result: MigrationResult| {
             lease_contract.and_then(|lease| {
-                batch
-                    .schedule_migrate_wasm_no_reply(&lease, MigrateMsg {}, self.new_code_id)
-                    .map(|_| batch)
-                    .map_err(Into::into)
+                self.leases_left -= 1;
+                result
+                    .try_add_msgs(|msgs| {
+                        msgs.schedule_migrate_wasm_no_reply(&lease, MigrateMsg {}, self.new_code_id)
+                            .map_err(Into::into)
+                    })
+                    .map(|mut res| {
+                        if self.leases_left == NbInstances::MIN {
+                            res.last_instance = Some(lease);
+                        }
+                        res
+                    })
             })
         };
 
-        self.may_batch = self.may_batch.and_then(op);
+        self.may_result = self.may_result.and_then(op);
         self
     }
 }
 
-impl TryFrom<MigrateBatch> for Batch {
+impl TryFrom<MigrateBatch> for MigrationResult {
     type Error = ContractError;
     fn try_from(this: MigrateBatch) -> Result<Self, Self::Error> {
-        this.may_batch
+        this.may_result
     }
 }
 
 #[cfg(test)]
 mod test {
```

Comment:
No tests for `max_leases = 0`. Either such should be added or `NonZero*` variant should be used.

Reply:
added a test scenario at `paging()`